src/backend/common/MemoryManager.hpp

```diff
@@ -233,74 +248,81 @@ class MemoryManager
 
     void unlock(void *ptr, bool user_unlock)
     {
+        using std::unique_ptr;
+        using std::function;
         // Shortcut for empty arrays
         if (!ptr) return;
 
-        lock_guard_t lock(this->memory_mutex);
-        memory_info& current = this->getCurrentMemoryInfo();
+        // Frees the pointer outside the lock.
+        unique_ptr<void, function<void(void*)>> freed_ptr(nullptr,
+                                                          [this](void* p) {
+                                                              this->nativeFree(p);
+                                                          });
```

Comment:
Move this into the class definition of MemoryManager ? It could be useful elsewhere.